modules/internal/ChapelRange.chpl

```diff
@@ -924,8 +940,8 @@ operator :(r: range(?), type t: range(?)) {
     tmp._aligned = r.aligned;
   }
 
-  tmp._low = r.low: tmp.intIdxType;
-  tmp._high = r.high: tmp.intIdxType;
+  tmp._low = if r.hasLowBound() then r._low else 1: tmp.intIdxType;
```

Comment:
Is there a difference between this cast and a call to `chpl__intToIdx` (where we might want to use the latter?).

Reply:
Yes, there is a difference.  As background, internally, a range stores its bounds as some form of integer, which is the `intIdxType`.  So the `_low` and `_high` fields in a range are this internal integer type.  The best motivation for why we do this is enums where for `enum color {red = 0xff0000, green=0x00ff00, blue = 0x0000ff}, we want to store the bounds of the range `red..blue` as something like `0..2` internally rather than, say, `0xff0000..0x0000ff`.

Here, we're assigning to `tmp._low`, so we're setting this internal representation.  Specifically, to either `r._low` or `1` depending on whether `r._low` is meaningful or not (where the `1` is fairly arbitrary and was chosen because it's "what we were doing before this change").   If we were assigning to `r.low`, we would use `intToIdx` would would convert the internal integer type to the user-facing `idxType`.  But we can't actually assign to `r.low` because it's a read-only query implemented using a paren-less function.

The reason the cast is required is that `r` may have an idxType/intIdxType of `int(64)`, say, while `tmp` may have an idxType/intIdxType of `int(32)`, requiring a downcast to make it fit.

I think the old code was actually incorrect in that if the range were an enum, we'd be trying to cast the enum bound to an integer, which isn't supported for abstract enums like `enum size {small, medium, large};`.  So I should really add a test to verify this and lock the fix in.

Having typed all of this up, it also occurs to me that since the `1` value is arguably bogus here, we could probably simply write this as `tmp._low = r._low: tmp.intIdxType;` without any loss of accuracy.

(also, in a late breaking "code cleanup", I broke this which is the version you reviewed which is what made me realize what the cast was doing. I've just pushed an update which fixes it to:

```chapel
tmp._low = (if r.hasLowBound() then r._low else 1): tmp.intIdxType;
```

although having typed the above I think I'll add the test of enum range casting and try the simplified form.

Reply:
Glad my confusion could help indirectly move things along!

Reply:
@dlongnecke-cray : I added more cast tests to improve confidence that the right thing was happening in bounded, unbounded, enum, and strided cases, which helped me get those lines into the appropriate final form (and also revealed a pre-existing bug in strided range casts).  It would probably make sense for you to review the tests added in https://github.com/chapel-lang/chapel/pull/17532/commits/b482e4d01ae1a02a957951b7723506aa2002fc69 (if not the code changes themselves) to see whether things behave as you'd expect.

Reply:
The changes seem reasonable to me.